### ISSUE TITLE
how to store oci json?

### ISSUE THREAD

**Issue by allisonkarlitskaya** [allisonkarlitskaya/composefs_experiments#11]
We currently support images and streams, which can each contain references to (only) blobs in `objects/`.  It might be useful to have a new type of file:

 - has its entire content inline.  zstd compressed?  Probably doesn't make a big difference, honestly...
 - despite having its entire content inline, it can contain out-of-band references to other items in the repository which are not directly included in its content
 - can refer to those items in ways other than by object-id.  For example, can refer to a `streams/by-sha256/abc...`.

The obvious usecase here is the various JSON files we find inside of OCI containers.  When we get those out of the repository we want to have the exact same thing we put in (like a splitstream), but we also want to make sure that so long as it's in the repository, we won't GC anything it refers to.

Another way of thinking about this is that we could add a feature to splitstreams to refer to external files without causing that data to become part of the content of the file itself.

There could be a lot of potential file formats, including "exploded" formats that store the references directly in the filesystem as a series of symlinks...  Or a binary format which lists the references up front.  Or a text format which lists them in text form, one per line (plus a header telling you how many bytes to skip to get to the data).  Or we could put the binary or text metadata at the end, instead of at the start.  We just need to pick something.

**Comment by allisonkarlitskaya** [comment]
After rolling this over in my head for a bit I'm increasingly convinced that it makes sense to represent JSON in the same format and namespace as layer tarballs.  They're just too similar, also in the way they get handled in the OCI spec.  They're all just blobs in the same namespace, right?

So all that's left is figuring out a way to encode the references in a way that's friendly to GC not needing to know how to parse JSON.  And I think a neat idea there might be to encode the references at the point of reference:  So if you have some JSON document like

```json
[
   "...",
   {
      "digest": "sha256:123456abcdef...",
      "...": "..."
   },
   "...",
]
```

Then in the splitstream this would be split into 3 blocks:
 - a normal "inline content" block up to the `"digest": `
 - a reference, which would expand to `"sha256:123456abcdef..."`
 - a normal "inline content" block from `,` to the end

Note: the entire JSON literal, including the quotes, is encoded in the reference.  This could be useful if we imagine a custom sort of JSON tree format which also supports external references...

This is getting kinda cheeky with our pretend claim to be OCI independent, however: we take a lot of their format into the core, including the quote characters and the `sha256:` prefix.  At least it's not an entire JSON parser?

We could also imagine a format where we include the entire `"digest": "sha256:` in the first inline chunk, the reference chunk expands *only* to the 64 byte hex sha256 hash value, and then the next inline chunk contains the closing quote: `",`.  This is a little bit more OCI-independent but it ends up splitting the literal, which makes encoding in some specialized tree format more conceptually muddy.

We could also imagine an approach where the entire JSON document is inline in a single chunk and the references are stored alongside it, purely for GC.

I'm going to start experimenting with the above three possibilities and see which one I end up liking more.

**Comment by allisonkarlitskaya** [comment]
I'm starting to think that splitstream wants to grow a header which contains an array of mappings from sha256 to fsverity values.

This would:
 - solve the issue at hand by giving us a list of sha256 references to pin down during GC
 - provide additional robustness guarantees for access at runtime: by including the fsverity digest of the underlying object in each stream which references a particular object we are also modifying the fsverity digest of *that* object in such a way that someone holding the original digest can be 100% certain about every bit they read while following links to it.

The idea is then if you have a config or a manifest and you follow a sha256 reference from it, you do a lookup in this table to (a) make sure the reference you're following is a valid reference, and (b) know its fs-verity digest, so you can measure the file when you open it.  With this approach, if the container runtime stored the final fsverity digest of the whole thing, we'd essentially be able to guarantee bit-for-bit that nothing got changed.  This is different from if we did lookups by sha256 alone, in which case we'd need to trust that the symlinks didn't get tampered/corrupted.

The way this works in standard OCI, of course, is that each object can have its sha256 content digest verified at time of access, but this is *slow*.  In our case, it would be pathologically slow.  JSON configs and manifests are tractable, but consider that the code behind `cfsctl oci create-image` essentially rips through the tarball layers without ever touching the data.  If we had to compute the sha256 of the original tar archive at time of access then we'd have to open every single file.  No thanks.

Splitstreams are zstd compressed, which means we have a potentially very-much-nondeterministic transformation involved in the chain of how the fsverity digests end up being produced.  That would limit the usefulness of these digests to the local machine.  We might also have an option to disable compression on JSON splitstreams, or indeed, do this as a separate file format.

**Comment by allisonkarlitskaya** [comment]
Note: trying to gain more determinism by disabling compression on JSON documents wouldn't be enough: we'd also need to disable compression on the tar streams, since their fsverity digests (computed on the compressed file) become part of the JSON file, even if they're uncompressed.

Note that zstd guarantees deterministic behaviour only given the same input, compression level, and version of zstd (ie: future versions would obviously want to have 'better' output).  That's quite reasonable, of course.

Another option would be introducing a 3rd kind of checksum: sha256 of the uncompressed splitstream.  This is sort of like OCI sha256 on uncompressed tarballs, so it's not that new of an idea.  We'd have to compute this ourselves as we read the splitstream, but at least it would be deterministic and wouldn't involve loading all of the data from the disk.  I kinda dislike this option, to be honest.

Maybe disabling compression in userspace that bad if we can get filesystem support.  We already know how a combination of filesystem-level encryption and fs-verity looks like (ext4):

> ext4 also supports encryption, which can be used simultaneously with
> fs-verity.  In this case, the plaintext data is verified rather than
> the ciphertext.  This is necessary in order to make the fs-verity file
> digest meaningful, since every file is encrypted differently.

Presumably some filesystem that supported compression (in the future?) would handle fs-verity digests of the data in the same way... That would get us our checksum over non-compressed data...

**Comment by allisonkarlitskaya** [comment]
On the sha256->fsverity mapping idea: imagine we create a new container image by adding a layer to an existing image.  In that case we'll reuse the layers from the previous image, but how do we know their fs-verity is correct?  We have that information in the slipstream from the previous image, so we're kinda sorta covered, but it feels very out-of-band.

But what about the case where we download a new container image that contains a layer that we already have on disk?  We want to generate a trusted fsverity digest for that new container image config and manifest but we need to know the fsverity of the original layers in order to do that.  All we have is a directory with a sha256 symlink to an object in the database.  We have no way to know that this link didn't get corrupt, unless we want to scan the entire thing to verify that it has the correct sha256 (which we probably *don't* want to do).

This argues towards putting fsverity digests in the manifests/configs themselves, and figuring out a way to make them stable.  @cgwalters you had some thoughts on this, I think?

**Comment by cgwalters** [comment]
> So all that's left is figuring out a way to encode the references in a way that's friendly to GC not needing to know how to parse JSON. 

This seems like a quite high level of effort to go to not parse JSON (or at a higher level, parse into OCI structures from JSON). I'm in favor of supporting usage of composefs outside of OCI but I think the practical way to do that is to make layers of libraries, which is what we have today in the composefs C library and its Rust crates right?

**Comment by cgwalters** [comment]
> This argues towards putting fsverity digests in the manifests/configs themselves, and figuring out a way to make them stable. @cgwalters you had some thoughts on this, I think?

Yes, stubbed those out in https://github.com/cgwalters/composefs-oci-experimental/blob/main/README.md#cfs-oci-native-annotations

As that README.md is getting to I think there's two different kinds of images: one whose manifest includes the annotations, and our role is to just verify them. The second kind of image is one without them, but we want to transform (or create a new copy of) the image with those annotations computed. The latter will commonly happen I think when we want to take a container image generated by an arbitrary other build process and "composefs-ify" it with a tool we ship.

**Comment by allisonkarlitskaya** [comment]
This more or less got resolved by 668b38695968b28a1e458805dfee3d84b38d06a4 and can be seen in action in 27505f88f9be96bc49054ada4e303edf97fd6cf6.
